aws-api/src/main/java/com/amplifyframework/api/aws/MutiAuthSubscriptionOperation.java

```diff
@@ -0,0 +1,255 @@
+                AmplifyException.TODO_RECOVERY_SUGGESTION
+            ));
+            return;
+        }
+        AppSyncGraphQLRequest<T> appSyncRequest = (AppSyncGraphQLRequest<T>) getRequest();
+        this.authTypes = MultiAuthModeStrategy.getInstance()
+                                              .authTypesFor(appSyncRequest.getModelSchema(),
+                                                            appSyncRequest.getAuthRuleOperation());
+    }
+
+    @NonNull
+    static <T> Builder<T> builder() {
+        return new Builder<>();
+    }
+
+    @Override
+    public synchronized void start() {
+        if (canceled.get()) {
+            onSubscriptionError.accept(new ApiException(
+                "Operation already canceled.", "Don't cancel the subscription before starting it!"
+            ));
+            return;
+        }
+        subscriptionFuture = executorService.submit(this::dispatchRequest);
+    }
+
+    private void dispatchRequest() {
+        LOG.debug("Processing subscription request: " + getRequest().getContent());
+        if (authTypes.hasNext()) {
+            AuthorizationType authorizationType = authTypes.next();
+            LOG.debug("Attempting to subscribe with " + authorizationType.name());
+            GraphQLRequest<T> request = getRequest();
+            // if the rule we're currently processing is an owner-based rule,
+            // then call the AuthRuleRequestDecorator to see if the owner needs to be
+            // added to the request.
+            if (authTypes.isOwnerBasedRule()) {
+                try {
+                    request = requestDecorator.decorate(request, authorizationType);
+                } catch (ApiException apiException) {
+                    LOG.warn("Unable to automatically add an owner to the request.", apiException);
+                    if (apiException instanceof ApiAuthException) {
+                        subscriptionFuture = executorService.submit(this::dispatchRequest);
+                    } else {
+                        emitErrorAndCancelSubscription(apiException);
+                    }
+                    return;
+                }
+            }
+            subscriptionEndpoint.requestSubscription(
+                request,
+                authorizationType,
+                subscriptionId -> {
+                    MutiAuthSubscriptionOperation.this.subscriptionId = subscriptionId;
+                    onSubscriptionStart.accept(subscriptionId);
+                },
+                item -> {
+                    if (item.hasErrors()) {
+                        for (GraphQLResponse.Error error : item.getErrors()) {
+                            if (error.getExtensions() != null &&
+                                UNAUTHORIZED_EXCEPTION.equals(error.getExtensions().get("errorType"))) {
+                                subscriptionFuture = executorService.submit(this::dispatchRequest);
```

Comment:
ah, so this is recursively trying out subscription with each auth rule until it succeeds?

Reply:
Correct. And it does so without adding to the current stack by simply queuing up the next attempt.